### PR TITLE
fix: correctly print counter-example for parallel execution

### PR DESCRIPTION
**Pull request by thalesmg**
When using parallel execution, the counter-example is a singleton list
with a single 2-tuple containing the sequential prefix and the
parallel executions.  But the code was expecting it to be an
enumerable.

```
  1) property broken ticket issuer (PropCheck.Test.BrokenTicketIssuerTest)
     test/broken_ticket_issuer_test.exs:8
     ** (Protocol.UndefinedError) protocol Enumerable not implemented for {[], [[{:set, {:var, 2}, {:call, PropCheck.Test.BrokenTicketIssuerTest, :inc, []}}, {:set, {:var, 3}, {:call, PropCheck.Test.BrokenTicketIssuerTest, :inc, []}}, {:set, {:var, 5}, {:call, PropCheck.Test.BrokenTicketIssuerTest, :inc, []}}], [{:set, {:var, 7}, {:call, PropCheck.Test.BrokenTicketIssuerTest, :inc, []}}, {:set, {:var, 8}, {:call, PropCheck.Test.BrokenTicketIssuerTest, :inc, []}}]]} of type Tuple. This protocol is implemented for the following type(s): Date.Range, File.Stream, Function, GenEvent.Stream, HashDict, HashSet, IO.Stream, List, Map, MapSet, Range, Stream
     code: property "broken ticket issuer", [:verbose] do
     stacktrace:
       (elixir 1.13.4) lib/enum.ex:1: Enumerable.impl_for!/1
       (elixir 1.13.4) lib/enum.ex:143: Enumerable.reduce/3
       (elixir 1.13.4) lib/enum.ex:4144: Enum.reduce/3
       (propcheck 1.4.2-dev) lib/properties.ex:300: PropCheck.Properties.counter_example_inspect/1
       (propcheck 1.4.2-dev) lib/properties.ex:249: PropCheck.Properties.handle_check_results/2
       test/broken_ticket_issuer_test.exs:8: (test)
```

Also, while running a test several times to confirm the fix, another
bug was found for the case when the parallel execution counter-example
has an empty sequential prefix.  In that case, the counter-example was
not being classified as a parallel one, and printed incorrectly.

```
  1) property broken ticket issuer (PropCheck.Test.BrokenTicketIssuerTest)
     test/broken_ticket_issuer_test.exs:8
     Property Elixir.PropCheck.Test.BrokenTicketIssuerTest.property broken ticket issuer() failed. Counter-Example is:
     [
       {[],
        [
          [
            {:set, {:var, 2},
             {:call, PropCheck.Test.BrokenTicketIssuerTest, :inc, []}},
            {:set, {:var, 3},
             {:call, PropCheck.Test.BrokenTicketIssuerTest, :inc, []}},
            {:set, {:var, 4},
             {:call, PropCheck.Test.BrokenTicketIssuerTest, :inc, []}},
            {:set, {:var, 5},
             {:call, PropCheck.Test.BrokenTicketIssuerTest, :inc, []}}
          ],
          [
            {:set, {:var, 8},
             {:call, PropCheck.Test.BrokenTicketIssuerTest, :inc, []}}
          ]
        ]}
     ]
```

### CODE REVIEW

**Comment by alfert** [comment]
Thanks a lot for identifying the bug and the fix.

**Comment by thalesmg** [comment]
Strange, locally I didn't get any errors in the tests, except for this one (which I get on `master` as well):

```
  1) test PropEr error is not stored (PropCheck.Test.CounterStrikeTest)
     test/counterstrike_test.exs:79
     Expected exception ExUnit.AssertionError but got UndefinedFunctionError (function :eunit_lib.fun_parent/1 is undefined or private)
     code: assert_raise ExUnit.AssertionError, ~r/cant_generate/, fn ->
     stacktrace:
       :eunit_lib.fun_parent(#Function<0.1423003/1 in PropCheck.Test.CounterStrikeTest.Helper."property cant_generate"/1>)
       (proper 1.4.0) /home/thales/dev/elixir/propcheck/deps/proper/src/proper_gen.erl:176: :proper_gen."-generate/3-lc$^0/1-0-"/1
       (proper 1.4.0) /home/thales/dev/elixir/propcheck/deps/proper/src/proper_gen.erl:176: :proper_gen.generate/3
       (proper 1.4.0) /home/thales/dev/elixir/propcheck/deps/proper/src/proper_gen.erl:136: :proper_gen.generate/1
       (proper 1.4.0) /home/thales/dev/elixir/propcheck/deps/proper/src/proper_gen.erl:123: :proper_gen.safe_generate/1
       (proper 1.4.0) /home/thales/dev/elixir/propcheck/deps/proper/src/proper.erl:1690: :proper.run/3
       (proper 1.4.0) /home/thales/dev/elixir/propcheck/deps/proper/src/proper.erl:1518: :proper.perform/7
       (proper 1.4.0) /home/thales/dev/elixir/propcheck/deps/proper/src/proper.erl:1257: :proper.inner_test/2
       (proper 1.4.0) /home/thales/dev/elixir/propcheck/deps/proper/src/proper.erl:1239: :proper.test/2
       (propcheck 1.4.2-dev) lib/properties.ex:188: PropCheck.Properties.execute_property/4
       test/counterstrike_test.exs:69: PropCheck.Test.CounterStrikeTest.Helper."property cant_generate"/1
       test/counterstrike_test.exs:85: (test)
```

But I couldn't find it in the CI tests :thinking: 

Perhaps it's the cover that's failing?

**Comment by thalesmg** [comment]
Oops, now I understood the failure.  The check for stored counter examples was seeing the failure of the new test I included.  :see_no_evil: 
(Although I'm still getting the `UndefinedError` for `:eunit_lib.fun_parent/1` :thinking: )

**Comment by thalesmg** [comment]
By the way, I noticed that `Reporter.pretty_print_counter_example_parallel/1` has a bit of overlap with `Reporter.print_report/2`.  However, the counter-example that the former has access to does not have the intermediate results and the result of the `run_parallel_commands`.

<details>
<summary> `Reporter.pretty_print_counter_example_parallel/1` </summary>

```
  1) property broken ticket issuer (PropCheck.Test.BrokenTicketIssuerTest)
     test/broken_ticket_issuer_test.exs:8
     Property Elixir.PropCheck.Test.BrokenTicketIssuerTest.property broken ticket issuer() failed. Counter-Example is:
     Sequential Start: 
        var1 = BrokenTicketIssuerTest.inc()
     
     Parallel Process 1: 
        var2 = BrokenTicketIssuerTest.inc()
     
     Parallel Process 2: 
        var3 = BrokenTicketIssuerTest.inc()
        var4 = BrokenTicketIssuerTest.inc()
     
     
     
     Counter example stored.
```
</details>

<details>
<summary>  `Reporter.print_report/2` </summary>

```
================================================================================
Concurrency Failure, we don't show the state :-/

Sequential commands:
   var1 = BrokenTicketIssuerTest.inc()
        # -> 0
        # Post state: 0


Process 1:
   var2 = BrokenTicketIssuerTest.inc()
        # -> 1


Process 2:
   var3 = BrokenTicketIssuerTest.inc()
        # -> 1
   var4 = BrokenTicketIssuerTest.inc()
        # -> 2



Result: :no_possible_interleaving
```

This is when using:
 
```elixir
(result == :ok)
      |> aggregate(command_names(cmds))
      |> when_fail((
        PropCheck.StateM.Reporter.print_report({seq_history, par_history, result}, cmds)
        IO.puts("Result: #{inspect(result)}")
      ))
```
</details>

Do you think it would make sense to unify them somehow?

**Comment by thalesmg** [comment]
I found another small bug when printing the report, for a case where there's no parallel history:

```
Shrinking .........(9 time(s))
{[
   {:set, {:var, 6}, {:call, TicketDispenserTest, :take_ticket, []}},
   {:set, {:var, 7}, {:call, TicketDispenserTest, :reset, []}}
 ], [[], []]}


  1) property issues unique tickets (TicketDispenserTest)
     test/ticket_dispenser_test.exs:9
     ** (CaseClauseError) no case clause matching: {:ok, []}
     code: property "issues unique tickets", [:verbose, numtests: 1_000] do
     stacktrace:
       (propcheck 1.4.2-dev) lib/statem/reporter.ex:156: PropCheck.StateM.Reporter.main_section/5
       (propcheck 1.4.2-dev) lib/statem/reporter.ex:93: PropCheck.StateM.Reporter.print_pretty_report/6
       test/ticket_dispenser_test.exs:20: anonymous fn/3 in TicketDispenserTest."property issues unique tickets"/1
       (stdlib 4.0.1) lists.erl:1442: :lists.foreach_1/2
       (proper 1.4.0) /home/thales/dev/elixir/propcheck/deps/proper/src/proper.erl:1963: :proper.shrink/4
       (proper 1.4.0) /home/thales/dev/elixir/propcheck/deps/proper/src/proper.erl:1397: :proper.get_result/3
       (proper 1.4.0) /home/thales/dev/elixir/propcheck/deps/proper/src/proper.erl:1261: :proper.inner_test/2
       (proper 1.4.0) /home/thales/dev/elixir/propcheck/deps/proper/src/proper.erl:1239: :proper.test/2
       (propcheck 1.4.2-dev) lib/properties.ex:183: PropCheck.Properties.execute_property/4
       test/ticket_dispenser_test.exs:9: (test)
```

I've added a fix in [this commit](https://github.com/alfert/propcheck/pull/214/commits/30b8d02cc27cd56b241ae7fd01ed540ff56c2e49), if you don't mind.

**Comment by alfert** [comment]
@thalesmg , sorry for my slow responses. But I need to take a deeper look into that and have time for that earliest at the weekend. Did not work on that code for quite some time ...

File: test/verify_storing_counterexamples.sh [hunk 1]
```diff
@@ -10,4 +10,21 @@ if [ $? -eq 0 ]; then
...
 fi
 
+# Verify that counterexamples for failing parallel tests are printed correctly
+out=$(mix test --include manual test/broken_ticket_issuer_test.exs --include will_fail:true)
+if [ $? -eq 0 ]; then
+    echo "$out" | cat
+    echo Test succeeded but should fail
+    exit 1
+else
+  if    ! grep -q "Counter-Example is:" <<< "$out" \
```
[the remark this answers is not in the document]
Comment (alfert) [reply test/verify_storing_counterexamples.sh:22]: TIL about here strings, which used. This is a feature not supported by the Bourne Shell (`/bin/sh`), but it runs on my Mac. Please change the script interpreter to `/bin/bash`

**Review by alfert — CHANGES_REQUESTED** [review]

File: test/verify_storing_counterexamples.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (thalesmg) [reply test/verify_storing_counterexamples.sh:22]: Indeed, they are valid in bash, and strangely enough I've just run the script with the original interpreter (`sh`) on my machine (Archlinux) and it worked.  :thinking: 

I've changed it here: https://github.com/alfert/propcheck/pull/214/commits/926ddf6952294419495ee38aa879cae6b8b11101

[the remark this answers is not in the document]
Comment (thalesmg) [reply test/verify_storing_counterexamples.sh:22]: Ha, of course it worked the same...  :facepalm: 

```sh
ͳ l $(which sh)
lrwxrwxrwx 1 root root 4 Jan  8  2022 /usr/bin/sh -> bash
```

[the remark this answers is not in the document]
Comment (alfert) [reply test/verify_storing_counterexamples.sh:22]: Similar things on macOS :-)

**Comment by alfert** [comment]
Thanks again for this bugfix!

**Comment by thalesmg** [comment]
Thanks for accepting the patch and for your work on the lib!  :beers:

**Pull request merged**